### PR TITLE
Add Azure Function App Support

### PR DESCRIPTION
**Pull request by ArchangelSDY**

### CODE REVIEW

**Comment by ArchangelSDY** [comment]
Integration test result:
http://jenkins-ci.westus2.cloudapp.azure.com/job/azure-webapp-plugin/job/dayshen/job/function/2/console

File: pom.xml [hunk 1]
```diff
@@ -238,6 +238,7 @@
                     <encoding>UTF-8</encoding>
                     <consoleOutput>true</consoleOutput>
                     <failsOnError>true</failsOnError>
+                    <resourceIncludes>**/*.properties,**/*.jelly,**/*.html</resourceIncludes>
```
Comment (juniwang) [inline pom.xml]: will jelly files be treated like ordinary html files? or any other standard for jelly files?

File: src/main/java/com/microsoft/jenkins/appservice/BaseDeploymentRecorder.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    protected final String azureCredentialsId;
+    protected final String resourceGroup;
+    protected final String appName;
+
+    @CheckForNull protected String filePath;
+    @CheckForNull protected String sourceDirectory;
+    @CheckForNull protected String targetDirectory;
+    protected boolean deployOnlyIfSuccessful;
+
+    protected BaseDeploymentRecorder(String azureCredentialsId, String resourceGroup, String appName) {
```
Comment (juniwang) [inline src/main/java/com/microsoft/jenkins/appservice/BaseDeploymentRecorder.java]: Will parameters without `final` fails in style check?

File: pom.xml (hunk 1 above)
Comment (ArchangelSDY) [inline pom.xml]: Actually for html/jelly files it only does some general checks such as white spaces, new line at end of file, etc.

File: src/main/java/com/microsoft/jenkins/appservice/BaseDeploymentRecorder.java (hunk 2 above)
Comment (ArchangelSDY) [inline src/main/java/com/microsoft/jenkins/appservice/BaseDeploymentRecorder.java]: Final parameters rule is disabled in current configuration. I copied these rules from azure-vm-agents-plugin and @ArieShout seems to have mentioned it in daily stand up weeks ago that this rule might be a bit annoying so just leave it disabled.

Comment (ArieShout) [inline src/main/java/com/microsoft/jenkins/appservice/BaseDeploymentRecorder.java]: I think we had a conclusion that we should have stricter rules. We just didn't update the rule :)

We enabled this rule in other plugins. Final parameters are enforced by default in some other languages such as Scala. Maybe we should enable this rule.

However, I would consider `final` methods, aka `DesignForExtension` rule as bad practice, as it is not friendly to tests, and not friendly when some other people want to extend the class behavior.

Comment (ArchangelSDY) [inline src/main/java/com/microsoft/jenkins/appservice/BaseDeploymentRecorder.java]: OK, I'll update the rule set.

@zackliu  Could you update the checkstyle configuration in azure-vm-agents plugin to make our rules consistent among all projects?

File: src/main/java/com/microsoft/jenkins/appservice/FunctionAppDeploymentCommandContext.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    public void setTargetDirectory(final String targetDirectory) {
+        this.targetDirectory = Util.fixNull(targetDirectory);
+    }
+
+    public void configure(
+            final Run<?, ?> run,
+            final FilePath workspace,
+            final TaskListener listener,
+            final FunctionApp app) throws AzureCloudException {
+        pubProfile = app.getPublishingProfile();
```
Comment (chenkennt) [inline src/main/java/com/microsoft/jenkins/appservice/FunctionAppDeploymentCommandContext.java]: Does function support slot deployment?

Comment (ArchangelSDY) [inline src/main/java/com/microsoft/jenkins/appservice/FunctionAppDeploymentCommandContext.java]: Slot for function is in preview and currently Azure SDK has no API for it.

**Pull request merged**